src/main/java/xfacthd/framedblocks/common/compat/jei/camo/CamoRecipeManagerPlugin.java

```diff
@@ -0,0 +1,215 @@
+            List<ItemStack> framedBlocks,
+            Ingredient firstInputs,
+            List<ItemStack> optionalOutputs
+    )
+    {
+        List<RecipeHolder<CraftingRecipe>> recipes = new ArrayList<>();
+        for (ItemStack framedBlock : framedBlocks)
+        {
+            // calculating the correct outputs here is impossible, leave them to be generated by
+            // CamoCraftingRecipeExtension.onDisplayedIngredientsUpdate
+            RecipeHolder<CraftingRecipe> singleRecipe = createRecipe(
+                    Ingredient.of(framedBlock),
+                    copyToolIngredient,
+                    firstInputs,
+                    Ingredient.EMPTY,
+                    optionalOutputs
+            );
+            recipes.add(singleRecipe);
+
+            if (camoCraftingHelper.isDoubleFramedBlock(framedBlock))
+            {
+                // calculating the correct outputs here is impossible, leave them to be generated by
+                // CamoCraftingRecipeExtension.onDisplayedIngredientsUpdate
+                RecipeHolder<CraftingRecipe> doubleRecipe = createRecipe(
+                        Ingredient.of(framedBlock),
+                        copyToolIngredient,
+                        firstInputs,
+                        camoCraftingHelper.getCamoExamplesIngredient(),
+                        optionalOutputs
+                );
+                recipes.add(doubleRecipe);
+            }
+        }
+
+        return recipes;
+    }
+
+    private List<RecipeHolder<CraftingRecipe>> createRepresentativeDoubleRecipes(
+            List<ItemStack> framedBlocks,
+            Ingredient firstInputs,
+            Ingredient secondInputs,
+            List<ItemStack> optionalOutputs
+    )
+    {
+        List<RecipeHolder<CraftingRecipe>> recipes = new ArrayList<>();
+        for (ItemStack framedBlock : framedBlocks)
+        {
+            // calculating the correct outputs here is impossible, leave them to be generated by
+            // CamoCraftingRecipeExtension.onDisplayedIngredientsUpdate
+            RecipeHolder<CraftingRecipe> doubleRecipe = createRecipe(
+                    Ingredient.of(framedBlock),
+                    copyToolIngredient,
+                    firstInputs,
+                    secondInputs,
+                    optionalOutputs
+            );
+            recipes.add(doubleRecipe);
+        }
+
+        return recipes;
+    }
+
+    private RecipeHolder<CraftingRecipe> createRecipe(
+            Ingredient frameStacks,
+            Ingredient copyTool,
+            Ingredient firstInputStacks,
+            Ingredient secondInputStacks,
+            List<ItemStack> results
+    )
+    {
+        JeiCamoCraftingRecipe recipe = new JeiCamoCraftingRecipe(frameStacks, copyTool, firstInputStacks, secondInputStacks, results);
+        ResourceLocation resourceLocation = ResourceLocation.fromNamespaceAndPath(FramedConstants.MOD_ID, "generated_recipe_" + generatedRecipeCount++);
```

Comment:
```suggestion
        ResourceLocation resourceLocation = Utils.rl("generated_recipe_" + generatedRecipeCount++);
```
Does the bookmarking care about this recipe ID? If it does, then the ID needs to be generated from the inputs/outputs as the registration order of the framed blocks is not stable (new blocks often get inserted in the middle of the type enum for sensible ordering in the creative tab).

Reply:
The recipe ID is only used for equality checks of RecipeHolder and in-memory checks. Persistent things like the bookmarks use the codec for serialization, which doesn't require stable IDs.

Reply:
I was wrong about this, the ID is also used to check if your bookmarks match a recipe that's being displayed.
I'll need to generate a more stable ID for that to work